File: MQTT/src/57/MQTT_SN_Gateway_Implementation.cpp
#include <iostream>
#include <map>
#include <vector>
#include <string>
#include <cstring>
#include <sys/socket.h>
#include <netinet/in.h>
#include <arpa/inet.h>
#include <unistd.h>
#include <mosquitto.h>

// MQTT-SN Message Types
#define MQTTSN_ADVERTISE     0x00
#define MQTTSN_SEARCHGW      0x01
#define MQTTSN_GWINFO        0x02
#define MQTTSN_CONNECT       0x04
#define MQTTSN_CONNACK       0x05
#define MQTTSN_REGISTER      0x0A
#define MQTTSN_REGACK        0x0B
#define MQTTSN_PUBLISH       0x0C
#define MQTTSN_SUBSCRIBE     0x12
#define MQTTSN_SUBACK        0x13
#define MQTTSN_PINGREQ       0x16
#define MQTTSN_PINGRESP      0x17
#define MQTTSN_DISCONNECT    0x18

// MQTT-SN Return Codes
#define MQTTSN_RC_ACCEPTED   0x00
#define MQTTSN_RC_CONGESTION 0x01
#define MQTTSN_RC_INVALID_TOPIC_ID 0x02
#define MQTTSN_RC_NOT_SUPPORTED 0x03

#pragma pack(push, 1)
struct MqttSnHeader {
    uint8_t length;
    uint8_t msg_type;
};

struct MqttSnConnect {
    MqttSnHeader header;
    uint8_t flags;
    uint8_t protocol_id;
    uint16_t duration;
    char client_id[23];
};

struct MqttSnRegister {
    MqttSnHeader header;
    uint16_t topic_id;
    uint16_t msg_id;
    char topic_name[256];
};

struct MqttSnPublish {
    MqttSnHeader header;
    uint8_t flags;
    uint16_t topic_id;
    uint16_t msg_id;
    char data[256];
};
#pragma pack(pop)

class MqttSnGateway {
private:
    int udp_socket;
    struct sockaddr_in gateway_addr;
    struct mosquitto* mqtt_client;
    
    std::map<uint16_t, std::string> topic_id_map;
    std::map<std::string, uint16_t> topic_name_map;
    std::map<std::string, sockaddr_in> client_addresses;
    uint16_t next_topic_id;
    
public:
    MqttSnGateway(const char* mqtt_host, int mqtt_port, int udp_port) 
        : next_topic_id(1) {
        
        // Initialize Mosquitto MQTT client
        mosquitto_lib_init();
        mqtt_client = mosquitto_new("mqttsn_gateway", true, this);
        
        if (mqtt_client == nullptr) {
            throw std::runtime_error("Failed to create MQTT client");
        }
        
        // Set MQTT callbacks
        mosquitto_message_callback_set(mqtt_client, on_mqtt_message_static);
        
        // Connect to MQTT broker
        if (mosquitto_connect(mqtt_client, mqtt_host, mqtt_port, 60) != MOSQ_ERR_SUCCESS) {
            throw std::runtime_error("Failed to connect to MQTT broker");
        }
        
        // Start MQTT loop in background
        mosquitto_loop_start(mqtt_client);
        
        // Create UDP socket for MQTT-SN
        udp_socket = socket(AF_INET, SOCK_DGRAM, 0);
        if (udp_socket < 0) {
            throw std::runtime_error("Failed to create UDP socket");
        }
        
        memset(&gateway_addr, 0, sizeof(gateway_addr));
        gateway_addr.sin_family = AF_INET;
        gateway_addr.sin_addr.s_addr = INADDR_ANY;
        gateway_addr.sin_port = htons(udp_port);
        
        if (bind(udp_socket, (struct sockaddr*)&gateway_addr, sizeof(gateway_addr)) < 0) {
            throw std::runtime_error("Failed to bind UDP socket");
        }
        
        std::cout << "MQTT-SN Gateway started on UDP port " << udp_port << std::endl;
        std::cout << "Connected to MQTT broker at " << mqtt_host << ":" << mqtt_port << std::endl;
    }
    
    ~MqttSnGateway() {
        mosquitto_loop_stop(mqtt_client, false);
        mosquitto_destroy(mqtt_client);
        mosquitto_lib_cleanup();
        close(udp_socket);
    }
    
    void run() {
        uint8_t buffer[1024];
        struct sockaddr_in client_addr;
        socklen_t addr_len = sizeof(client_addr);
        
        while (true) {
            ssize_t recv_len = recvfrom(udp_socket, buffer, sizeof(buffer), 0,
                                       (struct sockaddr*)&client_addr, &addr_len);
            
            if (recv_len > 0) {
                handle_mqttsn_message(buffer, recv_len, client_addr);
            }
        }
    }
    
private:
    void handle_mqttsn_message(uint8_t* buffer, ssize_t len, sockaddr_in& client_addr) {
        if (len < 2) return;
        
        MqttSnHeader* header = (MqttSnHeader*)buffer;
        
        std::cout << "Received MQTT-SN message type: 0x" << std::hex 
                  << (int)header->msg_type << std::dec << std::endl;
        
        switch (header->msg_type) {
            case MQTTSN_CONNECT:
                handle_connect(buffer, len, client_addr);
                break;
            case MQTTSN_REGISTER:
                handle_register(buffer, len, client_addr);
                break;
            case MQTTSN_PUBLISH:
                handle_publish(buffer, len, client_addr);
                break;
            case MQTTSN_SUBSCRIBE:
                handle_subscribe(buffer, len, client_addr);
                break;
            case MQTTSN_PINGREQ:
                handle_pingreq(client_addr);
                break;
            case MQTTSN_DISCONNECT:
                handle_disconnect(client_addr);
                break;
        }
    }
    
    void handle_connect(uint8_t* buffer, ssize_t len, sockaddr_in& client_addr) {
        MqttSnConnect* msg = (MqttSnConnect*)buffer;
        std::string client_id(msg->client_id, strnlen(msg->client_id, sizeof(msg->client_id)));
        
        client_addresses[client_id] = client_addr;
        
        // Send CONNACK
        uint8_t connack[3] = {3, MQTTSN_CONNACK, MQTTSN_RC_ACCEPTED};
        sendto(udp_socket, connack, sizeof(connack), 0,
               (struct sockaddr*)&client_addr, sizeof(client_addr));
        
        std::cout << "Client connected: " << client_id << std::endl;
    }
    
    void handle_register(uint8_t* buffer, ssize_t len, sockaddr_in& client_addr) {
        MqttSnRegister* msg = (MqttSnRegister*)buffer;
        std::string topic_name(msg->topic_name, len - 6);
        
        uint16_t topic_id = next_topic_id++;
        topic_id_map[topic_id] = topic_name;
        topic_name_map[topic_name] = topic_id;
        
        // Send REGACK
        uint8_t regack[7];
        regack[0] = 7;
        regack[1] = MQTTSN_REGACK;
        memcpy(&regack[2], &topic_id, 2);
        memcpy(&regack[4], &msg->msg_id, 2);
        regack[6] = MQTTSN_RC_ACCEPTED;
        
        sendto(udp_socket, regack, sizeof(regack), 0,
               (struct sockaddr*)&client_addr, sizeof(client_addr));
        
        std::cout << "Topic registered: " << topic_name << " -> ID " << topic_id << std::endl;
    }
    
    void handle_publish(uint8_t* buffer, ssize_t len, sockaddr_in& client_addr) {
        MqttSnPublish* msg = (MqttSnPublish*)buffer;
        uint16_t topic_id = ntohs(msg->topic_id);
        
        if (topic_id_map.find(topic_id) != topic_id_map.end()) {
            std::string topic = topic_id_map[topic_id];
            int data_len = len - 7;
            
            // Publish to MQTT broker
            mosquitto_publish(mqtt_client, nullptr, topic.c_str(),
                            data_len, msg->data, 0, false);
            
            std::cout << "Published to MQTT: " << topic << std::endl;
        }
    }
    
    void handle_subscribe(uint8_t* buffer, ssize_t len, sockaddr_in& client_addr) {
        // Extract topic name and subscribe to MQTT broker
        std::string topic((char*)buffer + 5, len - 5);
        
        mosquitto_subscribe(mqtt_client, nullptr, topic.c_str(), 0);
        
        // Send SUBACK
        uint8_t suback[8];
        suback[0] = 8;
        suback[1] = MQTTSN_SUBACK;
        suback[7] = MQTTSN_RC_ACCEPTED;
        
        sendto(udp_socket, suback, sizeof(suback), 0,
               (struct sockaddr*)&client_addr, sizeof(client_addr));
        
        std::cout << "Subscribed to topic: " << topic << std::endl;
    }
    
    void handle_pingreq(sockaddr_in& client_addr) {
        uint8_t pingresp[2] = {2, MQTTSN_PINGRESP};
        sendto(udp_socket, pingresp, sizeof(pingresp), 0,
               (struct sockaddr*)&client_addr, sizeof(client_addr));
    }
    
    void handle_disconnect(sockaddr_in& client_addr) {
        std::cout << "Client disconnected" << std::endl;
    }
    
    static void on_mqtt_message_static(struct mosquitto* mosq, void* obj,
                                      const struct mosquitto_message* msg) {
        ((MqttSnGateway*)obj)->on_mqtt_message(msg);
    }
    
    void on_mqtt_message(const struct mosquitto_message* msg) {
        // Forward MQTT message to MQTT-SN clients
        std::string topic(msg->topic);
        
        if (topic_name_map.find(topic) != topic_name_map.end()) {
            uint16_t topic_id = topic_name_map[topic];
            
            // Create MQTT-SN PUBLISH message
            std::vector<uint8_t> publish_msg(7 + msg->payloadlen);
            publish_msg[0] = publish_msg.size();
            publish_msg[1] = MQTTSN_PUBLISH;
            publish_msg[2] = 0; // flags
            memcpy(&publish_msg[3], &topic_id, 2);
            memcpy(&publish_msg[7], msg->payload, msg->payloadlen);
            
            // Broadcast to all registered clients
            for (auto& client : client_addresses) {
                sendto(udp_socket, publish_msg.data(), publish_msg.size(), 0,
                       (struct sockaddr*)&client.second, sizeof(client.second));
            }
            
            std::cout << "Forwarded MQTT message to MQTT-SN clients: " << topic << std::endl;
        }
    }
};

int main(int argc, char* argv[]) {
    try {
        MqttSnGateway gateway("localhost", 1883, 1884);
        gateway.run();
    } catch (const std::exception& e) {
        std::cerr << "Error: " << e.what() << std::endl;
        return 1;
    }
    
    return 0;
}

File: MQTT/src/57/MQTT_SN_Gateway_Implementation.rs
use std::collections::HashMap;
use std::net::SocketAddr;
use std::sync::Arc;
use tokio::net::UdpSocket;
use tokio::sync::Mutex;
use rumqttc::{AsyncClient, MqttOptions, QoS, Event, Packet};
use bytes::{BytesMut, Buf, BufMut};

// MQTT-SN Message Types
const MQTTSN_ADVERTISE: u8 = 0x00;
const MQTTSN_SEARCHGW: u8 = 0x01;
const MQTTSN_GWINFO: u8 = 0x02;
const MQTTSN_CONNECT: u8 = 0x04;
const MQTTSN_CONNACK: u8 = 0x05;
const MQTTSN_REGISTER: u8 = 0x0A;
const MQTTSN_REGACK: u8 = 0x0B;
const MQTTSN_PUBLISH: u8 = 0x0C;
const MQTTSN_PUBACK: u8 = 0x0D;
const MQTTSN_SUBSCRIBE: u8 = 0x12;
const MQTTSN_SUBACK: u8 = 0x13;
const MQTTSN_PINGREQ: u8 = 0x16;
const MQTTSN_PINGRESP: u8 = 0x17;
const MQTTSN_DISCONNECT: u8 = 0x18;

// Return Codes
const MQTTSN_RC_ACCEPTED: u8 = 0x00;
const MQTTSN_RC_CONGESTION: u8 = 0x01;
const MQTTSN_RC_INVALID_TOPIC_ID: u8 = 0x02;

#[derive(Debug, Clone)]
struct ClientInfo {
    address: SocketAddr,
    client_id: String,
}

struct MqttSnGateway {
    udp_socket: Arc<UdpSocket>,
    mqtt_client: AsyncClient,
    topic_id_map: Arc<Mutex<HashMap<u16, String>>>,
    topic_name_map: Arc<Mutex<HashMap<String, u16>>>,
    client_map: Arc<Mutex<HashMap<String, ClientInfo>>>,
    next_topic_id: Arc<Mutex<u16>>,
}

impl MqttSnGateway {
    async fn new(
        mqtt_host: &str,
        mqtt_port: u16,
        udp_port: u16,
    ) -> Result<Self, Box<dyn std::error::Error>> {
        // Setup MQTT client
        let mut mqtt_options = MqttOptions::new("mqttsn_gateway", mqtt_host, mqtt_port);
        mqtt_options.set_keep_alive(std::time::Duration::from_secs(60));

        let (mqtt_client, mut eventloop) = AsyncClient::new(mqtt_options, 10);

        // Spawn MQTT event loop
        tokio::spawn(async move {
            loop {
                match eventloop.poll().await {
                    Ok(event) => {
                        if let Event::Incoming(Packet::Publish(p)) = event {
                            println!("Received MQTT message on topic: {}", p.topic);
                        }
                    }
                    Err(e) => {
                        eprintln!("MQTT connection error: {:?}", e);
                        tokio::time::sleep(tokio::time::Duration::from_secs(1)).await;
                    }
                }
            }
        });

        // Setup UDP socket
        let udp_addr = format!("0.0.0.0:{}", udp_port);
        let udp_socket = UdpSocket::bind(&udp_addr).await?;

        println!("MQTT-SN Gateway started on UDP port {}", udp_port);
        println!("Connected to MQTT broker at {}:{}", mqtt_host, mqtt_port);

        Ok(Self {
            udp_socket: Arc::new(udp_socket),
            mqtt_client,
            topic_id_map: Arc::new(Mutex::new(HashMap::new())),
            topic_name_map: Arc::new(Mutex::new(HashMap::new())),
            client_map: Arc::new(Mutex::new(HashMap::new())),
            next_topic_id: Arc::new(Mutex::new(1)),
        })
    }

    async fn run(&self) -> Result<(), Box<dyn std::error::Error>> {
        let mut buf = vec![0u8; 1024];

        loop {
            let (len, addr) = self.udp_socket.recv_from(&mut buf).await?;
            
            if len >= 2 {
                let msg_type = buf[1];
                println!("Received MQTT-SN message type: 0x{:02X} from {}", msg_type, addr);
                
                self.handle_message(&buf[..len], addr).await;
            }
        }
    }

    async fn handle_message(&self, data: &[u8], addr: SocketAddr) {
        if data.len() < 2 {
            return;
        }

        let msg_type = data[1];

        match msg_type {
            MQTTSN_SEARCHGW => self.handle_searchgw(addr).await,
            MQTTSN_CONNECT => self.handle_connect(data, addr).await,
            MQTTSN_REGISTER => self.handle_register(data, addr).await,
            MQTTSN_PUBLISH => self.handle_publish(data, addr).await,
            MQTTSN_SUBSCRIBE => self.handle_subscribe(data, addr).await,
            MQTTSN_PINGREQ => self.handle_pingreq(addr).await,
            MQTTSN_DISCONNECT => self.handle_disconnect(addr).await,
            _ => println!("Unknown MQTT-SN message type: 0x{:02X}", msg_type),
        }
    }

    async fn handle_searchgw(&self, addr: SocketAddr) {
        // Send GWINFO
        let mut response = BytesMut::new();
        response.put_u8(3); // length
        response.put_u8(MQTTSN_GWINFO);
        response.put_u8(1); // gateway ID

        if let Err(e) = self.udp_socket.send_to(&response, addr).await {
            eprintln!("Failed to send GWINFO: {}", e);
        }
    }

    async fn handle_connect(&self, data: &[u8], addr: SocketAddr) {
        if data.len() < 6 {
            return;
        }

        let client_id_len = data[0] as usize - 6;
        let client_id = if data.len() >= 6 + client_id_len {
            String::from_utf8_lossy(&data[6..6 + client_id_len]).to_string()
        } else {
            "unknown".to_string()
        };

        // Store client info
        let client_info = ClientInfo {
            address: addr,
            client_id: client_id.clone(),
        };
        
        self.client_map.lock().await.insert(client_id.clone(), client_info);

        // Send CONNACK
        let mut response = BytesMut::new();
        response.put_u8(3); // length
        response.put_u8(MQTTSN_CONNACK);
        response.put_u8(MQTTSN_RC_ACCEPTED);

        if let Err(e) = self.udp_socket.send_to(&response, addr).await {
            eprintln!("Failed to send CONNACK: {}", e);
        }

        println!("Client connected: {} from {}", client_id, addr);
    }

    async fn handle_register(&self, data: &[u8], addr: SocketAddr) {
        if data.len() < 7 {
            return;
        }

        let mut cursor = std::io::Cursor::new(&data[2..]);
        let topic_id_from_client = cursor.get_u16();
        let msg_id = cursor.get_u16();
        
        let topic_name = String::from_utf8_lossy(&data[6..]).trim_end_matches('\0').to_string();

        // Allocate new topic ID
        let mut next_id = self.next_topic_id.lock().await;
        let topic_id = *next_id;
        *next_id += 1;
        drop(next_id);

        // Register topic
        self.topic_id_map.lock().await.insert(topic_id, topic_name.clone());
        self.topic_name_map.lock().await.insert(topic_name.clone(), topic_id);

        // Send REGACK
        let mut response = BytesMut::new();
        response.put_u8(7); // length
        response.put_u8(MQTTSN_REGACK);
        response.put_u16(topic_id);
        response.put_u16(msg_id);
        response.put_u8(MQTTSN_RC_ACCEPTED);

        if let Err(e) = self.udp_socket.send_to(&response, addr).await {
            eprintln!("Failed to send REGACK: {}", e);
        }

        println!("Topic registered: {} -> ID {}", topic_name, topic_id);
    }

    async fn handle_publish(&self, data: &[u8], addr: SocketAddr) {
        if data.len() < 8 {
            return;
        }

        let flags = data[2];
        let topic_id = u16::from_be_bytes([data[3], data[4]]);
        let msg_id = u16::from_be_bytes([data[5], data[6]]);
        let payload = &data[7..];

        let topic_map = self.topic_id_map.lock().await;
        if let Some(topic_name) = topic_map.get(&topic_id) {
            // Publish to MQTT broker
            if let Err(e) = self.mqtt_client
                .publish(topic_name, QoS::AtMostOnce, false, payload)
                .await
            {
                eprintln!("Failed to publish to MQTT: {}", e);
            } else {
                println!("Published to MQTT topic: {}", topic_name);
            }

            // Send PUBACK if QoS > 0
            if flags & 0x60 != 0 {
                let mut response = BytesMut::new();
                response.put_u8(7);
                response.put_u8(MQTTSN_PUBACK);
                response.put_u16(topic_id);
                response.put_u16(msg_id);
                response.put_u8(MQTTSN_RC_ACCEPTED);

                let _ = self.udp_socket.send_to(&response, addr).await;
            }
        }
    }

    async fn handle_subscribe(&self, data: &[u8], addr: SocketAddr) {
        if data.len() < 6 {
            return;
        }

        let msg_id = u16::from_be_bytes([data[3], data[4]]);
        let topic_name = String::from_utf8_lossy(&data[5..]).trim_end_matches('\0').to_string();

        // Subscribe to MQTT broker
        if let Err(e) = self.mqtt_client.subscribe(&topic_name, QoS::AtMostOnce).await {
            eprintln!("Failed to subscribe to MQTT topic: {}", e);
        } else {
            println!("Subscribed to MQTT topic: {}", topic_name);
        }

        // Allocate topic ID if not exists
        let mut name_map = self.topic_name_map.lock().await;
        let topic_id = if let Some(&id) = name_map.get(&topic_name) {
            id
        } else {
            let mut next_id = self.next_topic_id.lock().await;
            let id = *next_id;
            *next_id += 1;
            drop(next_id);

            self.topic_id_map.lock().await.insert(id, topic_name.clone());
            name_map.insert(topic_name.clone(), id);
            id
        };

        // Send SUBACK
        let mut response = BytesMut::new();
        response.put_u8(8);
        response.put_u8(MQTTSN_SUBACK);
        response.put_u8(0); // flags
        response.put_u16(topic_id);
        response.put_u16(msg_id);
        response.put_u8(MQTTSN_RC_ACCEPTED);

        if let Err(e) = self.udp_socket.send_to(&response, addr).await {
            eprintln!("Failed to send SUBACK: {}", e);
        }
    }

    async fn handle_pingreq(&self, addr: SocketAddr) {
        let mut response = BytesMut::new();
        response.put_u8(2);
        response.put_u8(MQTTSN_PINGRESP);

        let _ = self.udp_socket.send_to(&response, addr).await;
    }

    async fn handle_disconnect(&self, addr: SocketAddr) {
        println!("Client disconnected from {}", addr);
    }
}

#[tokio::main]
async fn main() -> Result<(), Box<dyn std::error::Error>> {
    let gateway = MqttSnGateway::new("localhost", 1883, 1884).await?;
    gateway.run().await?;
    Ok(())
}